visualize-admin/visualization-tool
# feat: Add a way to retrieve observations from multiple cubes

This PR adapts the application logic to support retrieving observations from multiple cubes at once (merging them along the way).

### GQL / SPARQL
- Removed `DataCubeObservations` query (replaced by `DataCubesObservations`)
- Observations queried from multiple cubes are now merged on the server-side by joinBy dimensions (assumed to be temporal dimensions of the same unit in the first iteration). This would need to be guaranteed by the front-end
- Removed unused variables (`rawObservations`, `sparql`) from observations query
- Introduced separate filter types per query (components, metadata, observations)
- Optimized hierarchy query to not re-fetch dimension values in case no filters were applied

---
## Context
This PR is a third part of preparations for merging the cubes. For the technical concept, see [this Notion document](https://www.notion.so/interactivethings/Combine-two-cubes-in-the-same-chart-d5711b414a7747e79edfc3cfe17d2c9b).

- [x] `DataCubesComponents`
- [x] `DataCubesMetadata`
- [x] `DataCubesObservations`

## How to test
1. Click around in the application and make sure nothing is broken 💥 
2. Open [GQL playground](https://visualization-tool-git-feat-multiple-cubes-observations-ixt1.vercel.app/api/graphql) and send e.g. the following query to see that observations from both cubes was fetched (and merged when encountering the same value from joinBy dimensions).

**Query**
```graphql
query DataCubesObservations(
  $sourceType: String!
  $sourceUrl: String!
  $locale: String!
  $filters: [DataCubeObservationFilter!]!
) {
  dataCubesObservations(
    sourceType: $sourceType
    sourceUrl: $sourceUrl
    locale: $locale
    filters: $filters
  )
}
```

**Variables**
```js
{
  "sourceType": "sparql",
  "sourceUrl": "https://lindas.admin.ch/query",
  "locale": "en",
  "filters": [
    {
      "iri": "https://energy.ld.admin.ch/sfoe/bfe_ogd84_einmalverguetung_fuer_photovoltaikanlagen/9",
      "latest": true,
      "filters": {
        "https://energy.ld.admin.ch/sfoe/bfe_ogd84_einmalverguetung_fuer_photovoltaikanlagen/Kanton": {
          "type": "single",
          "value": "https://ld.admin.ch/canton/1"
        }
      },
      "joinBy":"https://energy.ld.admin.ch/sfoe/bfe_ogd84_einmalverguetung_fuer_photovoltaikanlagen/Jahr"
    }
,
    {
      "iri": "https://energy.ld.admin.ch/sfoe/bfe_ogd18_gebaeudeprogramm_co2wirkung/4",
      "filters": {
        "https://energy.ld.admin.ch/sfoe/bfe_ogd18_gebaeudeprogramm_co2wirkung/region": {
          "type": "single",
          "value": "https://ld.admin.ch/country/CHE"
        },
        "https://energy.ld.admin.ch/sfoe/bfe_ogd18_gebaeudeprogramm_co2wirkung/massnahmenbereich": {
          "type": "single",
          "value": "https://energy.ld.admin.ch/sfoe/bfe_ogd18_gebaeudeprogramm_co2wirkung/ogd18_catalog/Haustechnik"
        }
      },
      "joinBy":"https://energy.ld.admin.ch/sfoe/bfe_ogd18_gebaeudeprogramm_co2wirkung/Jahr"
    } 
  ]
}
```
